src/web/lib/messagePicker.ts

```diff
@@ -0,0 +1,123 @@
+import { record } from '@root/src/web/browser/ophan/ophan';
+
+export type MaybeFC = React.FC | null;
+type ShowMessage = (meta?: any) => MaybeFC;
+
+export type CanShowResult = {
+	result: boolean;
+	meta?: any;
+};
+
+export type Candidate = {
+	id: string;
+	show: ShowMessage;
+	canShow: () => Promise<CanShowResult>;
+};
+
+export type CandidateConfig = {
+	candidate: Candidate;
+	timeoutMillis: number | null;
+};
+
+type CandidateConfigWithTimeout = CandidateConfig & {
+	cancelTimeout: () => void;
+};
+
+export type SlotConfig = {
+	candidates: CandidateConfig[];
+	name: string;
+};
+
+const recordMessageTimeoutInOphan = (messageId: string, slotName: string) =>
+	record({
+		component: `${slotName}-picker-timeout-dcr`,
+		value: messageId,
+	});
+
+const timeoutify = (
+	candidateConfig: CandidateConfig,
+	slotName: string,
+): CandidateConfigWithTimeout => {
+	let timer: number | undefined;
+
+	const canShow = (): Promise<CanShowResult> =>
+		new Promise((resolve) => {
+			if (candidateConfig.timeoutMillis) {
+				timer = window.setTimeout(() => {
+					recordMessageTimeoutInOphan(
+						candidateConfig.candidate.id,
+						slotName,
+					);
+					resolve({ result: false });
+				}, candidateConfig.timeoutMillis);
+			}
+
+			candidateConfig.candidate
+				.canShow()
+				.then((result) => resolve(result));
+		});
+
+	const cancelTimeout = () => timer && clearTimeout(timer);
+
+	return {
+		...candidateConfig,
+		candidate: {
+			...candidateConfig.candidate,
+			canShow,
+		},
+		cancelTimeout,
+	};
+};
+
+const clearAllTimeouts = (messages: CandidateConfigWithTimeout[]) =>
+	messages.map((m) => m.cancelTimeout());
+
+const defaultShow = () => null;
+
+type WinningMessage = {
+	idx: number;
```

Comment:
Would this be easier to use if it returned the winning candidate, rather than the index of the winning candidate? Typically it's nice to return information directly where possible, rather than the means to look it up. Apart from anything else, it removes the need to add code that checks that the index exists (or in general where the lookup fails).